src/primitives.rs

```diff
@@ -60,7 +66,53 @@ impl_versionize!(u64);
 impl_versionize!(f32);
 impl_versionize!(f64);
 impl_versionize!(char);
-impl_versionize!(String);
+
+impl Versionize for String {
+    #[inline]
+    fn serialize<W: std::io::Write>(
+        &self,
+        writer: &mut W,
+        version_map: &VersionMap,
+        app_version: u16,
+    ) -> VersionizeResult<()> {
+        // It is better to fail early at serialization time.
+        if self.len() > MAX_STRING_LEN {
+            return Err(VersionizeError::StringLength(self.len()));
+        }
+
+        self.len().serialize(writer, version_map, app_version)?;
+        writer
+            .write_all(self.as_bytes())
+            .map_err(|e| VersionizeError::Io(e.raw_os_error().unwrap_or(0)))?;
+        Ok(())
+    }
+
+    #[inline]
+    fn deserialize<R: std::io::Read>(
+        mut reader: &mut R,
+        version_map: &VersionMap,
+        app_version: u16,
+    ) -> VersionizeResult<Self> {
+        let len = usize::deserialize(&mut reader, version_map, app_version)?;
+        // Even if we fail in serialize, we still need to enforce this on the hot path
```

Comment:
This comment says that `// Even if we fail in serialize, ...`. Shouldn't it be `// Even if we succeed in serialize, ...`? Failing when serializing means that the string is to long or serializing the length and the string bytes to the snapshot failed. If these two failed and basically, the string serialization failed, should we concern about this case when deserializing?

Reply:
Failure in serialization because string is to len prevents the user from creating snapshots that violate the constraints. We need the check on the hot path in order to fail gracefully instead of SIGABRT/panic! when trying to allocate memory for deserializing a string/vec. For example the `len` might have been corrupted and now has a value of 200GB.